### PR TITLE
secp256k1: Introduce internal arith package.

### PR DESCRIPTION
**Pull request by davecgh**
**This is rebased on #3731**.

The package currently defines several generic constant time helpers and modulus-agnostic 512-bit multiplication and squaring in the main package and are shared by the scalar implementation and both field implementations.

The field implementations will eventually be split into separate packages in the future to prepare for architecture-specific backend selection.

In order to smooth that transition, this moves the aforementioned helpers and modulus-agnostic code into an internal `arith` package and updates all call sites to use them.

While the helpers and arithmetic are already well exercised by the upper layers that use them, testing the primitives directly helps ensure their correctness in isolation, so this also adds dedicated tests for all functions moved to the new internal package.

No functional changes are intended.  The move is purely organizational and preserves the existing constant time semantics as well as the existing arithmetic.

It is implemented via a series of commits to ease the review process. See the individual commit descriptions for details.

### CODE REVIEW

File: dcrec/secp256k1/curve.go [hunk 1]
```diff
@@ -957,140 +943,337 @@ func splitK(k *ModNScalar) (ModNScalar, ModNScalar) {
...
+	//   φ(n*P) = n*φ(P)
+	//
+	// This allows the second table to be obtained by applying φ to each entry
+	// of the first table instead of computing each odd multiple independently.
+	//
+	// NOTE: φ(x,y) = (βx,y).  The Jacobian z coordinates are the same, so this
+	// math goes through.
+	for i := 1; i <= oddMultiplesPerHalf; i++ {
+		phiTable[i].Set(&pTable[i])
+		phiTable[i].X.Mul(endoBeta)
```
Comment (jholdstock) [inline dcrec/secp256k1/curve.go]: This was previously Normalized.

File: dcrec/secp256k1/internal/arith/consttime.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+// ConstantTimeEq64 returns 1 if a == b or 0 otherwise in constant time.
+func ConstantTimeEq64(a, b uint64) uint32 {
+	t := a ^ b
+	return uint32(((t | -t) >> 63) ^ 1)
+}
+
+// ConstantTimeNotEq64 returns 1 if a != b or 0 otherwise in constant time.
+func ConstantTimeNotEq64(a, b uint64) uint64 {
```
Comment (jholdstock) [inline dcrec/secp256k1/internal/arith/consttime.go]: Should this return `uint32`? Seems like the odd one out in this file.

File: dcrec/secp256k1/internal/arith/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+the 512-bit multiplication and 512-bit squaring.
+
+## Installation and Updating
+
+This package is internal and therefore is neither directly installed nor needs
+to be manually updated.
+
+## License
+
+Package field10x26 is licensed under the [copyfree](http://copyfree.org) ISC
```
Comment (jholdstock) [inline dcrec/secp256k1/internal/arith/README.md]:
```suggestion
Package arith is licensed under the [copyfree](http://copyfree.org) ISC
```

File: dcrec/secp256k1/internal/arith/arith_test.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+		a:    "100000000000000000000000000000000",
+	}, {
+		name: "max uint192",
+		a:    "ffffffffffffffffffffffffffffffffffffffffffffffff",
+	}, {
+		name: "2^192",
+		a:    "1000000000000000000000000000000000000000000000000",
+	}, {
+		name: "2^255",
+		a:    "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
```
Comment (jholdstock) [inline dcrec/secp256k1/internal/arith/arith_test.go]:
```suggestion
		a:    "8000000000000000000000000000000000000000000000000000000000000000",
```

File: dcrec/secp256k1/internal/arith/arith_test.go [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+	}
+	return val
+}
+
+// mustBigToUint256 converts a [big.Int] to an array of 4 uint64s representing a
+// 256-bit little-endian value.  It will panic if the big integer is larger than
+// the maximum 256-bit value.  This is only provided for the hard-coded
+// constants so errors in the source code can be detected. It will only (and
+// must only) be called with hard-coded values.
+func mustBigToUint256(v *big.Int) [4]uint64 {
```
Comment (jholdstock) [inline dcrec/secp256k1/internal/arith/arith_test.go:33]: > It will only (and must only) be called with hard-coded values.

There are a couple of places where this is called with values from `randBig`

File: dcrec/secp256k1/curve_test.go [hunk 6]
```diff
@@ -714,21 +749,35 @@ func TestNAF(t *testing.T) {
 		in:   "a2e79d200f27f2360fba57619936159b",
 	}}
 
+	// Add tests for the first small values since they collectively exercise
+	// every possible residue modulo 2^w, where w is the window width, multiple
+	// times.  Then append the extra deterministic tests for edge cases.
+	const smallValues = 2*(1<<wNAFWidth) + 1
+	tests := make([]wnafTest, 0, smallValues+len(extraTests))
+	for i := 0; i < smallValues; i++ {
+		tests = append(tests, struct {
```
Comment (jholdstock) [inline dcrec/secp256k1/curve_test.go]:
```suggestion
		tests = append(tests, wnafTest{
```

File: dcrec/secp256k1/curve.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: That's not part of this PR.  I guess technically it is because it's part of the entire chain, but it's part of a previous one.

File: dcrec/secp256k1/internal/arith/consttime.go (hunk 2 above)
Comment (davecgh) [reply dcrec/secp256k1/internal/arith/consttime.go]: It was an intentional change since it's dealing with uint64 parameters, similar to `ConstantTimeSelect64`.  However, I see that `ConstantTimeEq64` is returning a `uint32` and that later ends up getting returned from other funcs that themselves return `uint32`, so it should probably stay `uint32` to avoid downcasting (though it honestly doesn't matter since they'll only ever by 0 or 1).

So, I agree it's better to keep them consistent.  Updated to return `uint32` and updated the callers accordingly.

File: dcrec/secp256k1/internal/arith/README.md (hunk 3 above)
Comment (davecgh) [reply dcrec/secp256k1/internal/arith/README.md]: Fixed.

File: dcrec/secp256k1/internal/arith/arith_test.go (hunk 4 above)
Comment (davecgh) [reply dcrec/secp256k1/internal/arith/arith_test.go]: Copied that one wrong.  Good call.  Updated.

File: dcrec/secp256k1/internal/arith/arith_test.go (hunk 5 above)
Comment (davecgh) [reply dcrec/secp256k1/internal/arith/arith_test.go:33]: True.  Modified to more accurately call out:

>// the maximum 256-bit value.  This is only provided for the hard-coded
> // constants and randomly-generated values that are known to be in range, so
> // errors in the source code can be detected.

File: dcrec/secp256k1/curve_test.go (hunk 6 above)
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Also not a part of this PR, but I'll update it in the other one.

File: dcrec/secp256k1/curve.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Also, it doesn't **really** need to be normalized because `Mul` produces magnitude one, but since the add and double funcs are documented as such, I updated it in the relevant PR to be extra safe.

File: dcrec/secp256k1/curve_test.go (hunk 6 above)
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go]: Updated in #3740.

**Review by jholdstock — APPROVED** [review]
